### PR TITLE
Automatically associate imported WP.com accounts

### LINKED ISSUE #17: Automatically associate migrated WP.com accounts
We should automatically associate migrated accounts with WP.com accounts for the purposes of Jetpack SSO. Association by email address should work.

Reported by @ryanmarkel

### PR DESCRIPTION
**Pull request by joshbetz**
Associates imported WP.com accounts by email. This ensures Jetpack SSO works
automatically, as expected.

Fixes #17

### CODE REVIEW

**Comment by joshbetz** [comment]
cc @ryanmarkel

**Comment by emrikol** [comment]
Wat.  Amazing.  This is only a one-liner?!

File: vip-go-wpcom-compat.php [hunk 1]
```diff
@@ -9,3 +9,12 @@
...
 require_once __DIR__ . '/wpcom-shortcodes.php';
+
+/**
+ * Jetpack SSO: Match WP.com accounts by email
+ *
+ * This ensures user accounts that have been imported from WordPress.com
+ * are still associated with the same WP.com account for the purpose
+ * of Jetpack SSO.
+ */
+add_filter( 'jetpack_sso_match_by_email', '__return_true', 9999 );
```
Comment (emrikol) [inline vip-go-wpcom-compat.php]: Instead of `9999` could we use `PHP_INT_MAX`?

**Comment by emrikol** [comment]
@joshbetz Is this the best place for this?  I was thinking that we need to clean the plugin up some, maybe make an `inc/` folder, and a `filters.php` (or something) file to clean things up a bit.  Thoughts?

Also, we should update the readme with this info, possibly in a `## Filters` section

**Comment by mjangda** [comment]
We presumably also need to force enable Jetpack SSO. Also, what testing do we need to do to make sure this works as expected?

**Comment by ryanmarkel** [comment]
Does this completely eliminate the connection process for users trying to do so? Or do they still need to go through the approval process?

**Comment by joshbetz** [comment]
We should test to be sure, but I think this makes the connection process unnecessary.

**Comment by joshbetz** [comment]
Noting this documentation https://developer.jetpack.com/hooks/jetpack_sso_match_by_email/

```
(bool) Should we link the local account to an account on WordPress.com using the same email address. Default to false.
```

**Comment by joshbetz** [comment]
@ryanmarkel I added this on a test site and enabled Jetpack SSO. I confirmed my account was not connected and logging in via SSO still works.

screen shot 2018-11-05 at 12 53 40 pm

**Comment by joshbetz** [comment]
By taking advantage of the `jetpack_active_modules` filter, we ensure the SSO module is enabled and that is reflected in the UI:

screen shot 2018-11-05 at 1 23 54 pm

**Comment by emrikol** [comment]
> @ryanmarkel I added this on a test site and enabled Jetpack SSO. I confirmed my account was not connected and logging in via SSO still works.

Did it log you in directly, or force you to go through the connection process?  How different was it from the WPCOM remote login flow?

> screen shot 2018-11-05 at 12 53 40 pm

Do we think this message will cause issues with less technical users who aren't expecting it?  Who might be confused why their WordPress.com account isn't "connected" anymore?  Especially if we lift and shift without telling them ;)

**Comment by joshbetz** [comment]
> Did it log you in directly, or force you to go through the connection process? How different was it from the WPCOM remote login flow?

It was the same flow. My account was still not connected after logging in.

> Do we think this message will cause issues with less technical users who aren't expecting it? Who might be confused why their WordPress.com account isn't "connected" anymore?

Yep, good point. We may want to hide this message.

**Comment by joshbetz** [comment]
> Do we think this message will cause issues with less technical users who aren't expecting it? Who might be confused why their WordPress.com account isn't "connected" anymore?

I think we're just going to have to deal with this banner because there are a couple problems:
1) While they don't need to go through the Jetpack connection flow in order for SSO to work (which is what we care about here), Jetpack doesn't technically consider them to be "connected."
2) There's not a straightforward way to hide this. The `is_user_connected()` function that's called is just checking for the existence of an access token. We could add a filter, but we'd be disabling connecting all users because we don't know based only on their email address if they're a WP.com user. It would only be a problem for new users that we don't import from WordPress.com, but still not great.

https://github.com/Automattic/jetpack/blob/master/class.jetpack.php#L1706-L1716

**Comment by joshbetz** [comment]
Related, we need to make sure there's no login loop if we're enforcing SSO: https://github.com/Automattic/vip-support/issues/64

**Comment by joshbetz** [comment]
Related, we need to make sure there's no login loop if we're enforcing SSO: https://github.com/Automattic/vip-support/issues/64

**Pull request merged**